### DISCUSSION TITLE
RLS UPDATE not working

### DISCUSSION THREAD

**Discussion by atumas-bananamilk** [supabase/supabase#7838] [Questions] [answered]
Hi,

I was following the tutorials on [egghead.io](egghead.io) (Tutorial 20 / 27), but the **UPDATE** on "profile" table isn't working.
Everything else works but "stripe_customer" doesn't get updated.
![Screenshot 2022-07-21 at 18 09 52](https://user-images.githubusercontent.com/22228431/180248972-ffdc505a-0bd1-4ed9-af80-334d797767ff.png)

Here's my RLS. I tried so many different variations, the only way it works is if i put "true" on the whole RLS.
![Screenshot 2022-07-21 at 18 02 44](https://user-images.githubusercontent.com/22228431/180248330-43cf2bb3-c6c6-4826-9f30-267e9d939709.png)

Here's my code:

`const stripe = new Stripe(process.env.STRIPE_SECRET!, {apiVersion: '2020-08-27',})`
`const customer = await stripe.customers.create({ email: req.body.record.email, })`

`console.log("ID: ", req.body.record.id)  // ---> returns 6a6e6f19-7f1b-49f4-9c62-c5d00871fc6b`
`console.log("CUSTOMER: ", customer.id) // ---> returns cus_M6ClB2RnePg8dH`

`await supabase.from('profile').update({ stripe_customer: customer.id, }, { returning: 'minimal', }).eq('id', req.body.record.id) // ---> this line isn't working because of RLS`

Can someone help?

**Comment by GaryAustin1** [comment]
If you set your RLS to just be the anon user (with true) does it work?  If so then you don't have a logged in user when you make that call.

**Reply by atumas-bananamilk** [reply]
It works with this yes.
![Screenshot 2022-07-21 at 19 04 17](https://user-images.githubusercontent.com/22228431/180262462-f29532f3-3f6e-4182-b9a5-770a5591e8ac.png)
![Screenshot 2022-07-21 at 19 14 02](https://user-images.githubusercontent.com/22228431/180262934-74b5ff9d-8810-403e-a255-30de56787260.png)


But I can see my user logged in all the time (I log the user on Navbar), but for some reason it's not enriched with data from "profile"table.

![Screenshot 2022-07-21 at 19 13 07](https://user-images.githubusercontent.com/22228431/180262990-2343f1d4-5f08-4cf8-a507-863ede5f53bc.png)

Code for enriching the user with "profile" data:

![Screenshot 2022-07-21 at 19 15 25](https://user-images.githubusercontent.com/22228431/180263233-c2735b9a-a90a-4e8a-b4f6-0743b9df7f1a.png)

**Reply by GaryAustin1** [reply]
If anon worked, then you are not really logged in, or your jwt is not getting set in the headers when you make your call.
I'm out of time to look at your code though for now.

**Reply by atumas-bananamilk** [reply]
It's the same code as in the tutorial. Nothing special. Thanks anyways. No idea how the user is "not logged in" still.

**Reply by GaryAustin1** [reply]
Can you see the network request for the update and look at the authorization header.  It should have a jwt much longer than the apiKey jwt.  If it is the same then that points back to user is not logged in.

There is no SB RLS, just Postgres RLS, besides the auth.uid() type functions.

For anon RLS to work, than means your jwt has the anon role in it (the default anon key).

The code you linked below is also 8 months old.  I don't know about your egghead tutorial as the link goes nowhere.

You can have a user id from a user object and still not have a valid session jwt (which RLS uses).   Which is probably why you can get the correct row (you have a filter on the user id) but the RLS is looking at the jwt.

**Reply by atumas-bananamilk** [reply]
- Network request for this update (`await supabase.from('profile').update({ stripe_customer: customer.id, }, { returning: 'minimal', }).eq('id', req.body.record.id`) is done on the server, not sure how to see it
- By RLS I mean Postgres RLS
- Any idea how my JWT got the role of "anon" after working for quite some time properly?
- Yes the tutorial is 7 months old, my code was written yesterday by following the tutorial:
https://egghead.io/courses/build-a-saas-product-with-next-js-supabase-and-stripe-61f2bc20
- No idea how to fix that

**Reply by GaryAustin1** [reply]
My GUESS on why it worked for a while, is it was after signUp or signIn and then the jwt or session expired and is not getting refreshed server side.  But that would by default only be an hour after signIn.

Edit:  Normally though the code would try to use the expired jwt I would think, and get an error, but I don't know how this code is keeping jwt refreshed.

Anon just means there is not a session and so the code uses the default key as the jwt for the call.

**Reply by GaryAustin1** [reply]
Not sure what your end goal is, but Supabase has been putting alot of time into this repository:
https://github.com/supabase-community/auth-helpers
For dealing with serverside/client auth functionality.  Does not address stripe directly...

**Comment by silentworks** [comment]
I think you are missing a rule in the **USING expression** section of the policy you have setup.

**Reply by atumas-bananamilk** [reply]
It was like this:

![Screenshot 2022-07-21 at 19 20 37](https://user-images.githubusercontent.com/22228431/180264269-ae24a3c6-850e-40bd-9500-a6d98f520d07.png)

**Reply by silentworks** [reply]
Add the same to the **WITH CHECK expression** field, both values are required for UPDATE operation to work.

**Reply by atumas-bananamilk** [reply]
Figured it out. Check my answer. Thanks!

**Comment by atumas-bananamilk** [comment]
I've just tested the same exact project made by someone else.
Everything works but the same error occurs.
"stripe_customer" isn't getting updated.

https://github.com/arzivre/saas-next-supabase

 I think at this point it's a bug within Supabase RLS.

**Comment by atumas-bananamilk** [comment, marked as answer]
I've figured this out.

The problem was that Supabase was doing `UPDATE` through "Function Hook" = Supabase server, but **not the user**.
That's why Supabase didn't have an `auth.uid()` and the way to solve this was just to continue watching the tutorials (Tutorial 22 / 27).

In the tutorial, they initialize another Supabase client for all backend services to use by **bypassing RLS**. All backend services operate server-side so it's all secure.

And it uses the service_role key:
![Screenshot 2022-07-22 at 02 54 03](https://user-images.githubusercontent.com/22228431/180333824-572a461e-02e4-4be8-9ebb-d4863a0a5a0b.png)

Here's a screenshot from the tutorial:
![Screenshot 2022-07-22 at 02 46 23](https://user-images.githubusercontent.com/22228431/180333566-8d8476cc-edf8-408b-9a0b-d03c73af7f04.png)
